src/main/java/com/cburch/logisim/std/memory/DualportRam.java

```diff
@@ -0,0 +1,423 @@
+  }
+
+  public boolean reset(CircuitState state, Instance instance) {
+    final var ret = (RamState) instance.getData(state);
+    if (ret == null) return true;
+    final var contents = ret.getContents();
+    if (instance.getAttributeValue(DualportRamAttributes.ATTR_TYPE).equals(DualportRamAttributes.VOLATILE)) {
+      contents.condClear();
+    }
+    return false;
+  }
+
+  @Override
+  public Bounds getOffsetBounds(AttributeSet attrs) {
+    return DualportRamAppearance.getBounds(attrs);
+  }
+
+  public MemContents getContents(InstanceState ramState) {
+    return getState(ramState).getContents();
+  }
+
+  @Override
+  MemState getState(Instance instance, CircuitState state) {
+    return getState(state.getInstanceState(instance));
+  }
+
+  @Override
+  MemState getState(InstanceState state) {
+    var ret = (RamState) state.getData();
+    if (ret == null) {
+      final var instance = state.getInstance();
+      ret = new RamState(instance, getNewContents(instance.getAttributeSet()), new MemListener(instance));
+      state.setData(ret);
+    } else {
+      ret.setRam(state.getInstance());
+    }
+    return ret;
+  }
+
+  @Override
+  protected void instanceAttributeChanged(Instance instance, Attribute<?> attr) {
+    super.instanceAttributeChanged(instance, attr);
+    if ((attr == Mem.DATA_ATTR)
+        || (attr == Mem.ADDR_ATTR)
+        || (attr == StdAttr.TRIGGER)
+        || (attr == DualportRamAttributes.ATTR_ByteEnables)
+        || (attr == StdAttr.APPEARANCE)
+        || (attr == Mem.LINE_ATTR)
+        || (attr == DualportRamAttributes.CLEAR_PIN)
+        || (attr == Mem.ENABLES_ATTR)) {
+      instance.recomputeBounds();
+      configurePorts(instance);
+    }
+  }
+
+  @Override
+  public void paintInstance(InstancePainter painter) {
+    if (DualportRamAppearance.classicAppearance(painter.getAttributeSet())) {
+      DualportRamAppearance.drawRamClassic(painter);
+    } else {
+      DualportRamAppearance.drawRamEvolution(painter);
+    }
+  }
+
+  @Override
+  public void propagate(InstanceState state) {
+    final var attrs = state.getAttributeSet();
+    final var myState = (RamState) getState(state);
+
+    // first we check the clear pin
+    if (attrs.getValue(DualportRamAttributes.CLEAR_PIN)) {
+      final var clearValue = state.getPortValue(DualportRamAppearance.getClrIndex(0, attrs));
+      if (clearValue.equals(Value.TRUE)) {
+        myState.getContents().clear();
+        final var dataBits = state.getAttributeValue(DATA_ATTR);
+
+        for (var i = 0; i < DualportRamAppearance.getNrDataOutPorts(attrs); i++) {
+          final var portVal = Value.createKnown(dataBits, 0);
+          state.setPort(DualportRamAppearance.getDataOutIndex(i, attrs), portVal, DELAY);
+        }
+
+        return;
+      }
+    }
+
+    // next we get the address and the mem value currently stored
+    final var addrValue = state.getPortValue(DualportRamAppearance.getAddrIndex(0, attrs));
```

Comment:
For consistency, consider `addrValue1` for the first memory port. Proceed analogously with other variables related to the first port.